app/questionnaire/completeness.py

```diff
@@ -0,0 +1,161 @@
+
+        group_states = [
+            self.get_state_for_group(group) for group in section['groups']
+        ]
+
+        if not group_states:
+            return self.NOT_STARTED
+
+        if all(state == self.SKIPPED for state in group_states):
+            return self.SKIPPED
+
+        if all(state in [self.COMPLETED, self.SKIPPED] for state in group_states):
+            return self.COMPLETED
+
+        if any(state in [self.STARTED, self.COMPLETED] for state in group_states):
+            return self.STARTED
+
+        return self.NOT_STARTED
+
+    def get_state_for_group(self, group, group_instance=None):
+        if isinstance(group, str):
+            # lookup group by group ID
+            group = self.schema.get_group(group)
+
+        if self._should_skip_group(group):
+            return self.SKIPPED
+
+        block_states = [
+            state for location, state in
+            self._get_block_states_for_group(group, group_instance)
+        ]
+
+        if self.routing_path and not block_states:
+            # the questionnaire has been started but none of the blocks for this group
+            # are on the current routing path
+            return self.SKIPPED
+
+        elif any(block_states):
+            if all(block_states):
+                return self.COMPLETED
+
+            return self.STARTED
+
+        return self.NOT_STARTED
+
+    def get_first_incomplete_location_in_survey(self):
+        incomplete_locations = (
+            self.get_first_incomplete_location_in_section(section)
+            for section in self._get_all_question_sections()
+        )
+        return next(filter(None, incomplete_locations), None)
+
+    def get_first_incomplete_location_in_section(self, section):
+        incomplete_locations = (
+            self.get_first_incomplete_location_in_group(group)
+            for group in section['groups']
+        )
+        return next(filter(None, incomplete_locations), None)
+
+    def get_first_incomplete_location_in_group(self, group, group_instance=None):
+        if self._should_skip_group(group):
+            return
+
+        incomplete_locations = (
+            location for location, is_complete in
+            self._get_block_states_for_group(
+                group, group_instance=group_instance)
+            if not is_complete
+        )
+        return next(incomplete_locations, None)
+
+    def is_block_valid_on_path(self, location, block):
+        if block['type'] == 'Question':
+            # Only question blocks are checked for completeness
+            return location in self.routing_path
+
+    def _get_block_states_for_group(self, group, group_instance=None):
+        if group_instance is not None:
+            current_instance = max_instance = group_instance
+        else:
+            max_instance = self._calculate_no_of_repeats(group) - 1
+            current_instance = 0
+
+        while current_instance <= max_instance:
```

Comment:
Does this need to be a `while` loop?
we have all the information upfront so it could just be a `for`?

Reply:
Not sure - I think might be clearer as a `while` loop in this instance. Doing this as a for loop would be:

`for current_instance in range(start_instance, max_instance + 1):`

That feels less readable to me.